### DISCUSSION TITLE
Questions about wrapper partial function

### DISCUSSION THREAD

**Discussion by ShanZeHu** [ptitSeb/box64#772] [Q&A] [answered]
Dear developer, Hello, I am a software engineering student named Mixky now very interested in your open source project and is engaged in this part of wrapper related research, is currently exploring the impact of wrapper on the overall performance of box64, I wrote a few common test library functions from X86 to ARM architecture, I want to wrapper in I want to test the program runtime with the wrapper function enabled or disabled respectively, because the translation function needs to be enabled without the wrapper native library. I have encountered some errors in the process, so I would like to ask you if this part of the wrapper function is provided with parameters that can be turned off or enabled separately, or how I can go about solving this problemo(╥﹏╥)o

**Comment by ShanZeHu** [comment]
![2023-05-09_08-57-29](https://user-images.githubusercontent.com/101883141/236967618-5a948dda-39a3-4d66-9374-30a566400aa1.png)

**Comment by ptitSeb** [comment]
Some libs cannot be "unwrapped" for now. Mainly "libc" and "pthread", they have to be used wrapped because some of the system function that they used internaly if emulated are not supported (the FS/GS tls segment handling).
So, a program that is staticaly linked with libc might not work because of that.

You want to compare performances with and without wrapping the libs right?

(and thank you for your interest in box64 project :) )

**Reply by ShanZeHu** [reply, marked as answer]
Yes, I was trying to compare the performance of the wrapper with and without the wrapper. I am currently making it possible to run only using the mock library by commenting all GO functions in src/library_list.h, and this is generating some errors. I'm not sure if I'm doing it right, so I'd like to ask you for your comparative work on the performance of the wrapper. My current research direction is to output a paper on optimizing the wrapper for the Box64 project, and I'd also like to ask you what clear directions I can take for optimization or how to go about it.
and thank you for your help :)

**Reply by ptitSeb** [reply]
You can simply use `BOX64_EMULATED_LIBS=XXX;YYY` to use emulated libs for XXX and YYY for example, it will be simpler.
But again, emulating `libc` and `libm` will not work.
Now, commenting the `GO(XXX)` in `library_list.h` will work, but you have to be sure that the x86_64 lib are available to box64. Also starting with glibc 2.35 (iirc), all libc, pthread, libm... are now fused in libc. So it might be usefull to use older version of glibc for this test.

**Reply by ShanZeHu** [reply]
Okay, I'll go ahead and try to follow this approach, I still have a lot of things I don't understand clearly enough, I'll try to familiarize myself with the whole project first, thanks for your help and answers

**Reply by ShanZeHu** [reply]
Dear ptiseb, after your guidance I was able to solve my problem, by debugging my test program I found that the IncRefCount function in the /librarian/library.c file was the key to the problem, then today I checked the original posting and found that it had been updated here, I synchronized it and found that my test program worked, the conclusion was that some small libraries such as libz, libpng with or without wrapper boost, the performance is about 7~10 times difference, thank you very much for your guidance and suggestions!

**Reply by ptitSeb** [reply]
It makes sense. libz or libpng will have tight loop of cpu intensive code to encode/decode, and sometime using hand optimized assembly code. This is were the wrapping will be the more efficiant!

**Reply by ShanZeHu** [reply]
I found another strange phenomenon, the improvement of wrapper for libz and libpng is really obvious, but compared to these two libraries which don't consume much cpu computing resources, the libjpeg and libfreetype library is more resource intensive because of the complex algorithms and conversions needed in its encoding and decoding process. However, in the process of testing this more resource-intensive library, I found that its runtime is slightly shorter without the wrapper. Here I give a set of jpeg test data:What do you think might cause this to happen? :(
![1683857625050](https://github.com/ptitSeb/box64/assets/101883141/a47ef0e8-4d3a-4855-ae68-15f2f0855a13)

**Reply by ShanZeHu** [reply]
And I found that libfreetype is also the case, the commonality between them is that libfreetype and libjpeg these two libraries will use cpu resources will be higher, so it will have regularity, or there is no possibility to call to some other libraries, GPU these parts, so I think this problem is more doubtful :(

**Reply by ptitSeb** [reply]
libjpeg is not using hand optimized asm function (but libturbojpeg is), the dynarec is pretty efficient on its code. Also, the wrapping of this lib is complex, as the structures needs to be analyzed before many call, and some changes needs to be made (handling of a jmpbuf inside the structure), making the wrapping less efficient.
it the case of libjpeg and freetype, the wrapping is there for conveniance: to avoid the end user the need to grab an x86_64 version of the lib to run their program.

**Reply by ShanZeHu** [reply]
Okay, thank you very much for your reply, the information you provided is very useful to me, I will make a partial summary and analysis of the work based on the existing information: )

**Reply by ShanZeHu** [reply]
Dear ptiseb, I currently have some ideas for this project myself, and one of the questions that bothers me is about the jmpbuf handling rules in the functions that you mentioned that WRAPPER does not manually optimize, does it have a different way of handling jmpbuf for each function or does it use a uniform rule for all functions to handle it? Or do I have a mistake in understanding this part, can you help me solve this problem?

**Reply by ptitSeb** [reply]
jumpbuff are wrapped and work fine when use directly. The iss ue is that the jmpbuf is setup in "arm world" but will be used in "x86" world. And that is not supported, that's why there is some special care in the jumpbuf structure.

**Reply by ShanZeHu** [reply]
Ok, thank you I understand. So what I want to know is, about the special treatment for the jmpbuf structure, is the jmp in each different function handled separately, and then is there room for this treatment to be optimized, and if so, I would like to base my related innovative research on this optimization.

**Comment by ShanZeHu** [comment]
Can you give me some advice on this  :)

**Reply by ptitSeb** [reply]
advice on what "this?

**Reply by ShanZeHu** [reply]
This refers to: for those library functions that use wrapper and have performance degradation due to jmp_buf processing (e.g. libjpeg, freetype, etc. that we mentioned before), I wonder if it is possible to optimize them for this part of jmpbuf processing, so as to solve this efficiency degradation

**Reply by ShanZeHu** [reply]
Is it possible to solve the problem that the use of wrapper leads to less efficient operation of certain library functions? If this problem is solvable, I would like to solve it, and then I would like to ask you to give me some possible directions to explore

**Reply by ShanZeHu** [reply]
Sorry, my previous statement may not have been clear :(

**Reply by ptitSeb** [reply]
Oh that. Well, I'm not sure. Wrapped should be more efficient in most cases. I'm not sure the constant wrapping/unwrapping of the structure really kills the performances, but I haven't mesured.
For correct image, the jmpbuf should not be use, so you can hack the source to remove the jmpbuf stuff and see if it's faster like that. Creating a jmpbuf should be quite fast, but the "sig" part of it might make it slower. Also, it's recreated every call iirc, so that also could be optimised (if the jmpbuf is the culprit)

**Comment by ShanZeHu** [comment]
I am most grateful for your detailed and insightful responses to my queries. They have significantly enhanced my understanding of wrapper function optimisation and the handling of the jmpbuf structure. I will certainly heed your suggestions for further experimentation and performance tuning in my project.

Once again, thank you for your generous assistance and for sharing your expertise！

**Comment by ShanZeHu** [comment]
Dear ptiseb，I have encountered a problem that I'm having trouble solving by myself and I was hoping you could assist me. I noticed that within the 'wrapped' directory, there are multiple library files of the same name but with different suffixes. In the **_private.h files, I found numerous GO function definitions, while the corresponding **.c files only seem to implement the wrapping of functions defined in GOM. If I wanted to add my own function wrappings, what criteria should I use to decide between using GO or GOM? Are there specific types of functions that need to be wrapped with GOM perhaps due to dynamically generated data that needs to be processed during runtime?

**Reply by ptitSeb** [reply]
`GO(XXX, ...)` are the automatic kind of wrapping. It's used for generic function that have simple argument and return values, like `int`, `char*` or structures that doesn't need alignments.
`GOM(XXX, ...)` are for using a `my_XXX(...)` function, because that function have some callback (function pointer that needs to be taken care of), or structures that needs some transformation. This is used when the automatic wrapping will not work.

You can have a look at this article: https://box86.org/2021/08/a-deep-dive-into-library-wrapping/ for more details about how the function wrapping works on box64.

**Reply by ShanZeHu** [reply]
The use of GO/GOM largely depends on the specifics of the function being wrapped. If the parameters and return values are relatively straightforward, GO would be the go-to choice. For weak functions, on the other hand, GOW would be more appropriate. The selection criteria mainly hinge on the particulars of the function being wrapped. It's largely up to my discretion; in essence, I should opt for GOW for more complex functions and GO for simpler ones. Is my understanding correct?

**Reply by ptitSeb** [reply]
No, the difference between GO and GOW is just the type of symbol. Use `nm -D` or something like that on the lib to see if the symbol is Weak or not.

**Reply by ShanZeHu** [reply]
Thank you for your clarification! This distinction between GO and GOW makes more sense to me now.

**Reply by ShanZeHu** [reply]
Previously, I've learned to utilize the nm -D command in order to distinguish between the use of GO and GOW. Currently, my understanding is that GOM should be applied for those functions involving callbacks and more complex arguments, with these functions typically prefixed by 'my_'. However, I'm uncertain about how to make this determination accurately for a given function. Should I refer to the original function definition to decide whether to use GOM or GO? Or is there a strategy to quickly judge if GO isn't functioning properly and hence needs to be switched to GOM? I'd appreciate any guidance on this.

**Reply by ptitSeb** [reply]
You need to use the original function definition for that, there is no other way.

**Reply by ShanZeHu** [reply]
Okay, I understand, thanks
